app/bundles/LeadBundle/EventListener/LeadSubscriber.php

```diff
@@ -274,45 +234,341 @@ public function onNotePostSave(Events\LeadNoteEvent $event)
+     */
+    public function onTimelineGenerate(Events\LeadTimelineEvent $event)
+    {
+        $eventTypes = [
+            'lead.create'       => 'mautic.lead.event.create',
+            'lead.identified'   => 'mautic.lead.event.identified',
+            'lead.ipadded'      => 'mautic.lead.event.ipadded',
+            'lead.utmtagsadded' => 'mautic.lead.event.utmtagsadded',
+            'lead.donotcontact' => 'mautic.lead.event.donotcontact'
+        ];
+
+        $filters = $event->getEventFilters();
+
+        foreach ($eventTypes as $type => $label) {
+            $name = $this->translator->trans($label);
+            $event->addEventType($type, $name);
+
+            if (!$event->isApplicable($type) || ($type != 'lead.utmtagsadded' && !empty($filters['search']))) {
+
+                continue;
+            }
+
+            switch ($type) {
+                case 'lead.create':
+                    $this->addTimelineDateCreatedEntry($event, $type, $name);
+                    break;
+
+                case 'lead.identified':
+                    $this->addTimelineDateIdentifiedEntry($event, $type, $name);
+                    break;
+
+                case 'lead.ipadded':
+                    $this->addTimelineIpAddressEntries($event, $type, $name);
+                    break;
+
+                case 'lead.utmtagsadded':
+                    $this->addTimelineUtmEntries($event, $type, $name);
+                    break;
+
+                case 'lead.donotcontact':
+                    $this->addTimelineDoNotContactEntries($event, $type, $name);
+                    break;
+            }
+        }
+    }
+
+    /**
+     * @param Events\LeadTimelineEvent $event
+     * @param                          $eventTypeKey
+     * @param                          $eventTypeName
+     */
+    protected function addTimelineIpAddressEntries(Events\LeadTimelineEvent $event, $eventTypeKey, $eventTypeName)
+    {
+        $lead = $event->getLead();
+        $rows = $this->auditLogModel->getRepository()->getLeadIpLogs($lead, $event->getQueryOptions());
+
+        if (!$event->isEngagementCount()) {
+            // Add to counter
+            $event->addToCounter($eventTypeKey, $rows);
+
+            // Add the entries to the event array
+            /** @var \Mautic\CoreBundle\Entity\AuditLog $row */
+            $ipAddresses = $lead->getIpAddresses()->toArray();
+
+            foreach ($rows['results'] as $row) {
+                if (!isset($ipAddresses[$row['ip_address']])) {
+
+                    continue;
+                }
+
+                $event->addEvent(
+                    [
+                        'event'           => $eventTypeKey,
+                        'eventLabel'      => $row['ip_address'],
+                        'eventType'       => $eventTypeName,
+                        'eventPriority'   => -1, // Usually an IP is added after another event
+                        'timestamp'       => $row['dateAdded'],
```

Comment:
Should this be `$row['date_added']`? Getting errors on dev and switching it to that fixed it.